src/theory/quantifiers/sygus/cegis_unif.cpp

```diff
@@ -0,0 +1,252 @@
+    Assert(d_enum_to_active_guard.find(e) != d_enum_to_active_guard.end());
+    Node g = d_enum_to_active_guard[e];
+    for (unsigned j = 0, size = enum_lems.size(); j < size; ++j)
+    {
+      enum_lems[j] = nm->mkNode(OR, g.negate(), enum_lems[j]);
+    }
+    lems.insert(lems.end(), enum_lems.begin(), enum_lems.end());
+  }
+  /* build candidate solution */
+  Assert(candidates.size() == 1);
+  Node vc = d_sygus_unif.constructSolution();
+  if (vc.isNull())
+  {
+    return false;
+  }
+  candidate_values.push_back(vc);
+  return true;
+}
+
+Node CegisUnif::purifyLemma(Node n,
+                            bool ensureConst,
+                            std::vector<Node>& model_guards,
+                            BoolNodePairMap& cache)
+{
+  unsigned i, size = n.getNumChildren();
+  Kind k = n.getKind();
+  bool fapp = k == APPLY_UF && size > 0 && n[0] == d_candidate;
+  Trace("cegis-unif-purify") << "PurifyLemma : " << n << "\n";
+  /* top level applications of function-to-synthsize should never be removed */
+  if (ensureConst || !fapp)
+  {
+    BoolNodePairMap::const_iterator it =
+        cache.find(BoolNodePair(ensureConst, n));
+    if (it != cache.end())
+    {
+      Trace("cegis-unif-purify") << "... already visited " << n << "\n";
+      return it->second;
+    }
+  }
+  /* Recurse */
+  bool childChanged = false;
+  std::vector<Node> children;
+  for (i = 0; i < size; ++i)
+  {
+    Node child = purifyLemma(n[i], ensureConst || fapp, model_guards, cache);
+    children.push_back(child);
+    childChanged = childChanged || child != n[i];
+  }
+  Node nb;
+  if (childChanged)
+  {
+    if (n.hasOperator())
+    {
+      children.insert(children.begin(), n.getOperator());
+    }
+    nb = NodeManager::currentNM()->mkNode(k, children);
+    Trace("cegis-unif-purify") << "PurifyFuncApp : transformed " << n
+                               << " into " << nb << "\n";
+  }
+  else
+  {
+    nb = n;
+  }
+  /* get model value if d_candidate applied over constants */
+  if (ensureConst && fapp)
+  {
+    Trace("cegis-unif-purify") << "PurifyFuncApp : candidate app " << nb
+                               << "\n";
+    for (i = 1; i < size; ++i)
+    {
+      if (!nb[i].isConst())
+      {
+        break;
+      }
+    }
+    if (i == size)
+    {
+      Node nv = d_parent->getModelValue(nb);
+      Trace("cegis-unif-purify") << "PurifyFuncApp : model value for " << nb
+                                 << " is " << nv << "\n";
+      /* test if different, then update model_guards */
+      if (nv != nb)
+      {
+        Trace("cegis-unif-purify") << "PurifyFuncApp : adding model eq\n";
+        model_guards.push_back(
+            NodeManager::currentNM()->mkNode(EQUAL, nv, nb).negate());
+        nb = nv;
+      }
+    }
+  }
+  /* top level applications of function-to-synthesize should not interfere with
+     internal applications being purified */
+  if (ensureConst || !fapp)
```

Comment:
Like the check at the top, I think this check can be removed.

We should then assert that:
Assert( !ensureConst || nb.isConst() );

Reply:
I removed the now spurious tests but I used a different assert. It would fail e.g. for ``(+ 1 2)``. What we need to guarantee is ``Assert(!ensureConst || !fapp || nb.isConst())``